include/cetl/any.hpp

```diff
@@ -26,69 +26,226 @@ class any;
-};
+    // Holds type-erased value destroyer. `nullptr` if storage has no value stored.
+    void (*value_destroyer_)(void* self) = nullptr;
 
-// Non-movable case.
-template <std::size_t Footprint, bool Copyable, std::size_t Alignment>
-struct base_move<Footprint, Copyable, false, Alignment> : base_copy<Footprint, Copyable, Alignment>
-{
 public:
-    constexpr base_move()                      = default;
-    base_move(base_move&&) noexcept            = delete;
-    base_move& operator=(base_move&&) noexcept = delete;
-};
+    // Holds type-erased value copyer. `nullptr` when copy operation is not supported.
+    void (*value_copier_)(const void* src, void* dst) = nullptr;
 
-// Copyable case.
-template <std::size_t Footprint, std::size_t Alignment>
-struct base_copy<Footprint, true, Alignment> : base_storage<Footprint, Alignment>
-{
-public:
-    constexpr base_copy()                  = default;
-    base_copy(const base_copy&)            = default;
-    base_copy& operator=(const base_copy&) = default;
-};
+    // Holds type-erased value mover. `nullptr` when move operation is not supported.
+    void (*value_mover_)(void* src, void* dst) = nullptr;
+
+    base_storage() = default;
+
+    CETL_NODISCARD void* get_raw_storage() noexcept
+    {
+        return static_cast<void*>(buffer_);
+    }
+
+    CETL_NODISCARD const void* get_raw_storage() const noexcept
+    {
+        return static_cast<const void*>(buffer_);
+    }
+
+    CETL_NODISCARD bool has_value() const noexcept
+    {
+        return nullptr != value_destroyer_;
+    }
+
+    template <typename Tp>
+    void make_handlers() noexcept
+    {
+        assert(nullptr == value_destroyer_);
+
+        value_destroyer_ = [](void* const storage) {
+            const auto ptr = static_cast<Tp*>(storage);
+            ptr->~Tp();
+        };
+    }
+
+    template <typename ValueType>
+    CETL_NODISCARD void* get_ptr() noexcept
+    {
+        if (!has_value())
+        {
+            return nullptr;
+        }
+
+        // TODO: Add RTTI check here.
+        return get_raw_storage();
+    }
 
+    template <typename ValueType>
+    CETL_NODISCARD const void* get_ptr() const noexcept
+    {
+        if (!has_value())
+        {
+            return nullptr;
+        }
+
+        // TODO: Add RTTI check here.
+        return get_raw_storage();
+    }
+
+    void copy_from(const base_storage& src)
+    {
+        if (src.has_value())
+        {
+            copy_handlers_from(src);
+            assert(nullptr != value_copier_);
+
+            value_copier_(src.get_raw_storage(), get_raw_storage());
+        }
+    }
+
+    void move_from(base_storage& src) noexcept
+    {
+        if (src.has_value())
+        {
+            copy_handlers_from(src);
+            assert(nullptr != value_mover_);
+
+            value_mover_(src.get_raw_storage(), get_raw_storage());
+
+            src.reset();
+        }
+    }
+
+    void copy_handlers_from(const base_storage& src) noexcept
```

Comment:
May I suggest that we avoid introducing unused data members by moving the type-erased copy and move handlers into the specializations of copy and move for which the corresponding operations are enabled? This would be similar to your telescoping `make_handlers` method where a derived class calls the base class.

This is not an optimization (nobody cares about the extra 8-16 bytes for the function pointers) but an attempt to use stronger typing and reduce the state space (no way to call a null function pointer if it's not there in the first place).

This change would affect:

- `copy_from` (only defined in `base_copy`; can be made private probably, see my comment about `swap`)
- `move_from` (only defined in `base_move`)
- `copy_handlers_from` (telescoping like `make_handlers`)
- `reset` (ditto)

Reply:
Please see new intermediate `base_handlers`.